File: crates/taflow-metrics/src/metrics/maximum_drawdown_test.rs
use super::maximum_drawdown::MaximumDrawdown;
use crate::{MetricInputKind, NanPolicy};

#[test]
fn uses_phantom_starting_wealth_and_returns_signed_decline() {
    let mut metric = MaximumDrawdown::new(NanPolicy::Omit).unwrap();
    metric.from_returns(&[]).unwrap();
    metric.extend(&[-0.2, 0.5, -0.5]).unwrap();

    // Wealth is 1 -> 0.8 -> 1.2 -> 0.6, so the deepest decline is -50%.
    assert_eq!(metric.value(), Some(-0.5));
    assert_eq!(metric.compute(), Some(-0.5));
    assert_eq!(metric.len(), 3);
}

#[test]
fn positive_first_return_does_not_create_a_drawdown() {
    let mut metric = MaximumDrawdown::new(NanPolicy::Omit).unwrap();
    metric.from_returns(&[]).unwrap();
    assert_eq!(metric.append(0.25).unwrap(), Some(0.0));
    assert!((metric.append(-0.1).unwrap().unwrap() + 0.1).abs() < 1e-15);
}

#[test]
fn all_semantic_input_modes_have_the_same_result() {
    let returns = [0.1, -0.2, 0.05];
    let expected = MaximumDrawdown::new(NanPolicy::Omit)
        .and_then(|mut state| {
            state.from_returns(&[])?;
            state.extend(&returns)?;
            Ok(state.compute())
        })
        .unwrap();

    let mut log_returns = MaximumDrawdown::new(NanPolicy::Omit).unwrap();
    log_returns.from_log_returns(&[]).unwrap();
    log_returns.extend(&returns.map(f64::ln_1p)).unwrap();

    let mut equity = MaximumDrawdown::new(NanPolicy::Omit).unwrap();
    equity.from_equity(&[]).unwrap();
    equity.extend(&[100.0, 110.0, 88.0, 92.4]).unwrap();

    let mut pnl = MaximumDrawdown::new(NanPolicy::Omit).unwrap();
    pnl.from_pnl(&[], 100.0).unwrap();
    pnl.extend(&[10.0, -22.0, 4.4]).unwrap();

    for actual in [log_returns.compute(), equity.compute(), pnl.compute()] {
        assert!((actual.unwrap() - expected.unwrap()).abs() < 1e-15);
    }
}

#[test]
fn reset_replay_and_chunking_preserve_lifecycle() {
    let returns = [0.1, -0.2, 0.05, -0.25, 0.1];
    let mut batch = MaximumDrawdown::new(NanPolicy::Omit).unwrap();
    batch.from_returns(&[]).unwrap();
    batch.extend(&returns).unwrap();

    let mut chunked = MaximumDrawdown::new(NanPolicy::Omit).unwrap();
    chunked.from_returns(&[]).unwrap();
    chunked.extend(&returns[..2]).unwrap();
    for value in &returns[2..] {
        chunked.append(*value).unwrap();
    }
    assert_eq!(chunked.compute(), batch.compute());

    chunked.reset();
    assert!(chunked.is_empty());
    assert_eq!(chunked.value(), None);
    chunked.extend(&returns).unwrap();
    assert_eq!(chunked.compute(), batch.compute());
}

#[test]
fn omitted_nan_does_not_advance_usable_length() {
    let mut metric = MaximumDrawdown::new(NanPolicy::Omit).unwrap();
    metric.from_returns(&[]).unwrap();
    metric.extend(&[0.1, f64::NAN, -0.25]).unwrap();
    assert_eq!(metric.len(), 2);
    assert_eq!(metric.value(), Some(-0.25));
}

#[test]
fn rejects_non_return_semantic_domains() {
    let mut unbound = MaximumDrawdown::new(NanPolicy::Omit).unwrap();
    assert!(unbound.append(0.01).is_err());
}

File: tests/metrics/maximum_drawdown_test.py
from __future__ import annotations
import empyrical
import numpy as np
import pytest
from taflow.metrics.maximum_drawdown import MaximumDrawdown

@pytest.mark.parametrize('returns', [np.array([0.01]), np.array([-0.2]), np.array([0.1, -0.2, 0.05]), np.zeros(32), np.linspace(-0.04, 0.05, 101), np.array([0.15, -0.1, -0.1, 0.25, -0.3, 0.1]), np.array([np.nan, 0.02, -0.03, np.nan, 0.01]), np.random.default_rng(20260811).normal(0.0004, 0.012, 513)])
def test_maximum_drawdown_matches_empyrical(returns: np.ndarray) -> None:
    actual = MaximumDrawdown().from_returns(returns).compute()
    expected = float(empyrical.max_drawdown(returns))
    assert actual == pytest.approx(expected, rel=1e-12, abs=1e-14)

def test_maximum_drawdown_input_methods_and_lifecycle_are_invariant() -> None:
    returns = np.array([0.1, -0.2, 0.05, -0.25, 0.1])
    expected = MaximumDrawdown().from_returns(returns).compute()
    equity = 100.0 * np.r_[1.0, np.cumprod(1.0 + returns)]
    pnl = np.diff(equity)
    assert MaximumDrawdown().from_equity(equity).compute() == pytest.approx(expected)
    assert MaximumDrawdown().from_pnl(pnl, initial_capital=100.0).compute() == pytest.approx(expected)
    assert MaximumDrawdown().from_log_returns(np.log1p(returns)).compute() == pytest.approx(expected)
    state = MaximumDrawdown().from_returns([])
    assert state.value is None
    assert state.append(returns[0]) is state
    assert state.extend(returns[1:3]) is state
    assert state.extend(returns[3:]) is state
    assert state.compute() == pytest.approx(expected)
    assert len(state) == len(returns)
    assert state.reset() is state
    assert len(state) == 0
    assert state.value is None
    assert state.extend(returns).compute() == pytest.approx(expected)

def test_maximum_drawdown_uses_phantom_wealth_one() -> None:
    assert MaximumDrawdown().from_returns([-0.2]).compute() == pytest.approx(-0.2)
    assert MaximumDrawdown().from_returns([0.2]).compute() == pytest.approx(0.0)

def test_maximum_drawdown_missing_and_invalid_values() -> None:
    assert len(MaximumDrawdown().from_returns([np.nan, 0.01])) == 1
    with pytest.raises(ValueError):
        MaximumDrawdown(nan_policy='raise').from_returns([np.nan])
    with pytest.raises(ValueError):
        MaximumDrawdown().from_returns([np.inf])
    with pytest.raises(ValueError):
        MaximumDrawdown().from_returns([-1.01])

def test_maximum_drawdown_requires_semantic_ingestion_before_append() -> None:
    with pytest.raises(ValueError):
        MaximumDrawdown().append(0.01)
